django/forms/widgets.py

```diff
@@ -588,6 +588,7 @@ def optgroups(self, name, value, attrs=None):
         """Return a list of optgroups for this widget."""
         groups = []
         has_selected = False
+        create_option = self.create_option
```

Comment:
I'm sceptical about this. Generating options is slow. I'm not convinced that avoiding the attribute lookup is the way to go with it though. A 4% speed-up is OK but the cost is obfuscate the code.   

We used to do this with JavaScript. Browsers were bad at navigating the scope hierarchy, so you'd create local copies of variables at the start of a function. It worked, a good speedup. But it was horrible. In the end they made browsers better, and I don't see it done these days (certainly not as a first-option). Ultimately it was the browser's issue. 

Same here. If this is an issue, it's Python's issue (IMO) — let them speed up attribute access. (If we're that worried about speed, there's a question about why we're using Python in the first place... but that gets us off-topic). 

`SelectDateWidget` is close to a worst-case example. I'm not convinced it's typical. Is it really the case that choices generation is the holdup in common scenarios…?

**And** if you are hitting issues here, the way forward (again IMO) is to cache choices generation, rather than bend the code the wrong way. I see us removing old micro-optimizations as unnecessary; not at all sure about adding more...